flink-core/src/main/java/org/apache/flink/core/execution/PipelineExecutor.java

```diff
@@ -41,5 +41,5 @@
 	 * @param configuration the {@link Configuration} with the required execution parameters
 	 * @return a {@link CompletableFuture} with the {@link JobClient} corresponding to the pipeline.
 	 */
-	CompletableFuture<JobClient> execute(final Pipeline pipeline, final Configuration configuration) throws Exception;
+	CompletableFuture<? extends JobClient> execute(final Pipeline pipeline, final Configuration configuration) throws Exception;
```

Comment:
@mxm Why was this changed from `CompletableFuture<? extends JobClient>`

Reply:
This was necessary to please Java generics when handling the derived JobClient class via futures. The alternative would have been to cast which we wanted to avoid. Since the class is marked `@Internal`, this should not be an issue?

Reply:
Yes, it's not a problem for the API. It just makes the internal signatures more wordy and I was surprised because I had to rebase some of my changes. 😅

I think the problem here is that Java in some situations can't type-check the code, if the `whenComplete()` would be removed it's happy to accept the result as a `CompletableFuture<JobClient>`.

Reply:
True, it makes things more verbose. If you find a way to remove it, go ahead. Perhaps sync with @tillrohrmann since he also reviewed the code and might have an idea.

Reply:
I think this change is not necessary and can be reverted. No explicit cast should be required if simply using an `thenApply(Function.identity())` with the right generics.